File: v2/internal/ai/sanitize/sanitize.go
// Package sanitize escapes adversarial content in source files before
// they are wrapped in XML blocks for AI context injection. This prevents
// prompt injection via crafted file contents.
//
// Author: Subash Karki
package sanitize

import "strings"

// xmlReplacer escapes XML-like tags that could break our context framing.
var xmlReplacer = strings.NewReplacer(
	"</codebase-context>", "&lt;/codebase-context&gt;",
	"</strategy-guidance>", "&lt;/strategy-guidance&gt;",
	"</phantom-context>", "&lt;/phantom-context&gt;",
	"</phantom-analysis>", "&lt;/phantom-analysis&gt;",
	"<system>", "&lt;system&gt;",
	"</system>", "&lt;/system&gt;",
	"<system-reminder>", "&lt;system-reminder&gt;",
	"</system-reminder>", "&lt;/system-reminder&gt;",
)

// XMLTags escapes XML-like tags in content that will be wrapped in XML blocks.
// This prevents prompt injection via source file contents that contain
// adversarial closing tags or system-level directives.
func XMLTags(content string) string {
	return xmlReplacer.Replace(content)
}

File: packages/server/src/utils/sanitize.ts
/**
 * Sanitize content before injecting into XML-tagged prompt blocks.
 * Prevents prompt injection via source file contents that contain
 * closing tags or system-level XML elements.
 * @author Subash Karki
 */

const DANGEROUS_PATTERNS: [RegExp, string][] = [
  [/<\/codebase-context>/gi, '&lt;/codebase-context&gt;'],
  [/<\/strategy-guidance>/gi, '&lt;/strategy-guidance&gt;'],
  [/<\/phantom-context>/gi, '&lt;/phantom-context&gt;'],
  [/<\/phantom-analysis>/gi, '&lt;/phantom-analysis&gt;'],
  [/<system>/gi, '&lt;system&gt;'],
  [/<\/system>/gi, '&lt;/system&gt;'],
  [/<system-reminder>/gi, '&lt;system-reminder&gt;'],
  [/<\/system-reminder>/gi, '&lt;/system-reminder&gt;'],
];

export const sanitizeXMLTags = (content: string): string => {
  let result = content;
  for (const [pattern, replacement] of DANGEROUS_PATTERNS) {
    result = result.replace(pattern, replacement);
  }
  return result;
};
